### ISSUE TITLE
How can I build the crate from cargo subcommand by using cargo library crate?

### ISSUE THREAD

**Issue by vityafx** [rust-lang/cargo#3916]
I am creating a cargo subcommand and I want to build the project if it was not built before. Also, it would be nice to get target build directory by using `cargo` crate as same as a way to build the crate. How can I do that?

**Comment by alexcrichton** [comment]
Hm I'm not quite sure I folllow, but maybe the source for [`cargo build`](https://github.com/rust-lang/cargo/blob/master/src/bin/build.rs) can help?

**Comment by vityafx** [comment]
Yes, though I can always see the code, I have asked about a minimal code for that, a short example.

**Comment by alexcrichton** [comment]
I'm sorry I don't really understand the question in that sense... can you expand on what you'd like to know?

**Comment by vityafx** [comment]
Yes, of course! So, here is what I want to do:

My subcommand (it is called `cargo cook` at this moment) gets the crate files, packages all of them into a container with compression and uploads somewhere. So, this subcommand always take build artifact from `target/release` directory (actually, the subcommand does not know where to take the artifact so this directory is specified in the config file but I want to delete it from there and to get this directory from `cargo` crate also) but here is the problem: suppose you called it once and it packaged your crate into `crate-0.1.1.tar.bzip2`. Then you found a mistake in the code, corrected it, updated version, rebuilt it but in debug, so you have `target/debug/crate` with 0.1.2 version but old one in `target/release` (version 0.1.1) and then you call my subcommand forgetting that fact and it packages old binary again. I want my subcommand to do almost the same as `cargo publish`: when called it builds the crate in release and then do what it is intended to do.

In short:
1. My subcommand was called.
2. My subcommand builds the crate in release mode and get's `target/release` directory from `cargo` crate (I want to remove this field from my configuration file).
3. My subcommands does what it want to do.

I hope it is more clear now, @alexcrichton.

**Comment by alexcrichton** [comment]
Hm in that case is Cargo-as-a-library needed? It seems like you could just call the `cargo` command line itself?

**Comment by vityafx** [comment]
Yes I could but this is not what I really want to do. And the `cargo` crate should have such docs even so :)

**Comment by alexcrichton** [comment]
Ah ok, yeah I definitely agree `cargo` itself could use far more API docs! In the meantime though you may want to explore `src/bin/build.rs` as you'll probably just do that that does, and you should be able to learn about the `target` directory from the `Config` passed in.

**Comment by vityafx** [comment]
@alexcrichton Yes, this is the only thing I can do I guess, so I will do it :)Thanks for the help anyway.

**Comment by alexcrichton** [comment]
Ok in that case I'm gonna close this, but feel free to ask any more questions if they arise!
